File: data_structures/trees/JavaScript/tree.js
// Basic Tree data structure implemented using JS

class Node {
  constructor(data, name) {
    this.name = name ? name : "generic tree node"; // <-- Name is an optional parameter that can be passed.  It is unnecessary for data structure but helpful for debugging
    this.data = data;
    this.children = [];
  }
  addChild(entry, name) {
    let child = new Node(entry, name);
    this.children.push(child);
    return child;
  }
}

class Tree {
  constructor(data, name) {
    this.root = new Node(data, name);
  }
  printNode(node) {
    console.log(`${node.name}: ${node.data}`);
    console.log(`Children of ${node.name}:`);
    node.children.forEach((childNode) => {
      console.log(`${childNode.name}: ${childNode.data}`);
    });
    console.log("");
    node.children.forEach((childNode) => {
      this.printNode(childNode);
    });
  }
  printTree() {
    this.printNode(this.root);
  }
  // Depth First Search and Breadth First Search Methods are added in the Algorithms folder
}

File: data_structures/trees/Python/tree.py
# Basic Tree data structure implemented using Python

class Node:
    def __init__(self, data, name):
        self.data = data
        self.name = name
        self.children = []
    def add_child(self, entry, name):
        child = Node(entry, name)
        self.children.append(child)
        return child

class Tree:
    def __init__(self, data, name):
        self.root = Node(data, name)
    def print_node(self, node):
        print(f"{node.name}: {node.data}")
        print(f"Children of {node.name}:")
        for child_node in node.children:
            print(f"{child_node.name}: {child_node.data}")
        print("")
        for child_node in node.children:
            self.print_node(child_node)
    def print_tree(self):
        self.print_node(self.root)